### PR TITLE
GraphQL based REST controllers

### PR DESCRIPTION
**Pull request by phyrog**
Exploring #149. Comments welcome.

Ping @eugenk, @ebolloff

### CODE REVIEW

**Comment by eugenk** [comment]
I like this. Well, there are a few things like naming up to discussion, but from the technical perspective, I really like this low-maintenance approach.

**Comment by phyrog** [comment]
Yep, I'm bad at naming things :smiley:

File: app/controllers/v3/dsl.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+      def query(query_string)
+        @query_string = query_string
+      end
+
+      def arguments(&block)
+        @arguments_proc = block
+      end
+
+      def context(&block)
```
Comment (phyrog) [inline app/controllers/v3/dsl.rb]: This is untested for now, as we don't have a controller yet that uses it.

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                      "example": null
+                    },
+                    {
+                      "type": "string",
+                      "example": "Ada Lovelace"
+                    }
+                  ]
+                },
+                "emailHash": {
+                  "type": "string"
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No example here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                  "$ref": "#/definitions/id"
+                },
+                "organizations": {
+                  "type": "array",
+                  "items": {
+                    "type": "object",
+                    "required": ["id"],
+                    "properties": {
+                      "id": {
+                        "type": "string"
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No example here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 4]
```diff
@@ -0,0 +1,139 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "title": "V3 Organizational Units",
+  "description": "V3 Organizational Unit Controller",
+  "definitions": {
+    "id": {
+      "type": "string",
+      "example": "ada",
+      "description": "The id of the user"
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: "The id of the organizational unit"

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                      "example": null
+                    },
+                    {
+                      "type": "string",
+                      "example": "Ontohub"
+                    }
+                  ]
+                },
+                "id": {
+                  "type": "string"
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No example here?

Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No description here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                },
+                "displayName": {
+                  "oneOf": [
+                    {
+                      "type": "null",
+                      "example": null
+                    },
+                    {
+                      "type": "string",
+                      "example": "Ontohub"
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No description here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                },
+                "description": {
+                  "oneOf": [
+                    {
+                      "type": "null",
+                      "example": null
+                    },
+                    {
+                      "type": "string",
+                      "example": "The Ontohub organization"
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No description here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json (hunk 3 above)
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No description here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                  ]
+                },
+                "emailHash": {
+                  "type": "string"
+                },
+                "id": {
+                  "$ref": "#/definitions/id"
+                },
+                "organizations": {
+                  "type": "array",
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No description here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                },
+                "displayName": {
+                  "oneOf": [
+                    {
+                      "type": "null",
+                      "example": null
+                    },
+                    {
+                      "type": "string",
+                      "example": "Ada Lovelace"
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: No description here?

File: spec/support/api/schemas/controllers/v3/organizational_units/get_show.json [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            }
+          ]
+        }
+      }
+    }
+  },
+  "links": [
+    {
+      "title": "Show",
+      "description": "Show the ontohub-backend version. Returns HTTP 200 (OK).",
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/organizational_units/get_show.json]: Copy/paste mistake.

File: spec/support/api/schemas/controllers/v3/version/get_show.json [hunk 11]
```diff
@@ -0,0 +1,53 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/version/get_show.json]: Upgrafe to draft-06

File: spec/support/api/schemas/controllers/v3/version/get_show.json [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      "title": "Show",
+      "description": "Show the ontohub-backend version. Returns HTTP 200 (OK).",
+      "rel": "self",
+      "method": "GET",
+      "mediaType": "application/json",
+      "href": "/v3/version",
+      "targetSchema": { "rel": "self" }
+    }
+  ]
+}
```
Comment (eugenk) [inline spec/support/api/schemas/controllers/v3/version/get_show.json]: Please add one schema per controller to the `spec/support/api/schemas/apidoc`

**Pull request merged**